### PR TITLE
core(driver): wait for LCP event

### PR DESCRIPTION
**Pull request by adamraine**
Addresses first action item of https://github.com/GoogleChrome/lighthouse/issues/11180

### CODE REVIEW

File: lighthouse-core/config/constants.js [hunk 1]
```diff
@@ -54,6 +54,7 @@ const throttling = {
 const defaultSettings = {
   output: 'json',
   maxWaitForFcp: 30 * 1000,
+  maxWaitForLcp: 40 * 1000,
```
Comment (adamraine) [inline lighthouse-core/config/constants.js:57]: This timing uses https://github.com/GoogleChrome/lighthouse/issues/11180#issuecomment-685931823 as a guideline but beyond that is pretty arbitrary. Is there a more definitive way to choose this timing?

File: lighthouse-core/gather/driver.js [hunk 2]
```diff
@@ -705,6 +709,36 @@ class Driver {
     };
   }
 
+  /**
+   * Returns a promise that resolve when a frame has a LCP.
```
Comment (connorjclark) [inline lighthouse-core/gather/driver.js:713]:
```suggestion
   * Returns a promise that resolves when a frame has a LCP.
```

File: lighthouse-core/gather/driver.js [hunk 3]
```diff
@@ -705,6 +709,36 @@ class Driver {
...
+    return this._waitForLifecycleEvent(
+      'largestContentfulPaint::Candidate',
+      pauseAfterLcpMs,
+      maxWaitForLcpMs,
+      LHError.errors.NO_LCP
+    );
+  }
+
+  /**
+   * Returns a promise that resolve when a frame has a FCP.
```
Comment (connorjclark) [inline lighthouse-core/gather/driver.js:728]:
```suggestion
   * Returns a promise that resolves when a frame has a FCP.
```

File: lighthouse-core/test/gather/driver-test.js [hunk 4]
```diff
@@ -702,7 +706,83 @@ describe('._waitForFcp', () => {
...
     expect(driver.off).toHaveBeenCalled();
-    await expect(waitPromise).rejects.toMatchObject({message: 'Wait for FCP canceled'});
+    await expect(waitPromise).rejects.toMatchObject(
+      {message: 'Wait for "firstContentfulPaint" event canceled'}
+    );
+  });
+});
+
+describe('._waitForLcp', () => {
+  it('should not resolve until LCP fires', async () => {
```
Comment (adamraine) [inline lighthouse-core/test/gather/driver-test.js:716]: These test cases are pretty much a copy/paste from `_waitForFcp` cases above.  Would it make more sense to combine these into a section for `_waitForLifecycleEvent` and add test case to differentiate FCP and LCP?

**Review by connorjclark — CHANGES_REQUESTED** [review]
Main concern: does this still fail the entire run given no LCP? I think Patrick was suggesting it should only impact the LCP audit, not the entire run.

**Comment by adamraine** [comment]
@connorjclark Doesn't the current behavior only fail audits dependent on LCP?

**Comment by patrickhulce** [comment]
assigning myself to remind me take a look :)

File: lighthouse-core/config/constants.js (hunk 1 above)
Comment (patrickhulce) [reply lighthouse-core/config/constants.js:57]: ah sorry @adamraine my comment was misleading. I don't think we actually need this to be a separate param compared to `maxWaitForLoad`, but I now that I'm looking at all the pieces, I kinda like the fine grained control and explicit it offers, so this WFM :)

File: lighthouse-core/gather/driver.js (hunk 2 above)
Comment (patrickhulce) [reply lighthouse-core/gather/driver.js:713]: heh, also in the original method where this was copied from :)

File: lighthouse-core/gather/driver.js [hunk 5]
```diff
@@ -1080,17 +1123,18 @@ class Driver {
    * possible workaround.
    * Resolves on the url of the loaded page, taking into account any redirects.
    * @param {string} url
-   * @param {{waitForFcp?: boolean, waitForLoad?: boolean, waitForNavigated?: boolean, passContext?: LH.Gatherer.PassContext}} options
+   * @param {{waitForFcp?: boolean, waitForLcp?: boolean, waitForLoad?: boolean, waitForNavigated?: boolean, passContext?: LH.Gatherer.PassContext}} options
    * @return {Promise<{finalUrl: string, timedOut: boolean}>}
    */
   async gotoURL(url, options = {}) {
     const waitForFcp = options.waitForFcp || false;
+    const waitForLcp = options.waitForLcp || false;
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver.js:1131]: I having trouble finding where we set this to `true`. Were there changes to gather-runner to enable this that I'm missing? :)

**Review by patrickhulce — COMMENTED** [review]
thanks @adamraine awesome job navigating this complicated area of LH! 🎉  :) taking a step back though I think we need to adjust our approach a bit.

I'm having trouble seeing this lifecycle event being fired by latest chrome, are we sure the candidates are fired over the protocol like this?

These are all the lifecycle events I saw of a navigation to example.com 

image
 
Once we get that sorted out, we'll also need to make sure that invalidations are fired as well because IIRC the root situation in the issues we're trying to solve is when LCP is invalidated and we need to wait for a new one (otherwise waiting for FCP should be enough). That unfortunately means we won't be able to share the same logic with FCP since we'll need to start waiting *again* if we receive an invalidation event and will probably need something more like `lcpQuietPeriod` rather than a `pauseAfterLcpMs` in the end. 

I know it's hairy but it's unlikely that we'll be getting a "this LCP is final" event from Chromium because the proposal for that event seems like it's going to be based on user input rather than any quiet heuristics :/

The best next step I think would be adding a smoke test that sets `lcpQuietWindow` to something high like 10s. Then the page invalidates LCP and only adds more content after 5s or so to make sure we wait for it. This hopefully helps ensure the impl is working as expected as you iterate on it.

**Comment by adamraine** [comment]
@patrickhulce thanks for the explanation.  I made the mistake of equating `largestContentfulPaint` trace events to a lifecycle event similar to FCP.  You are correct, there is not an LCP lifecycle event, are there any proposals to add one?

**Comment by patrickhulce** [comment]
> are there any proposals to add one?

I'm not aware of any but we can always ask DevTools team if it could be considered. I think the fact that there's no "final LCP" is tricky but we are already emitting FMP candidates there so it's at least consistent.

File: lighthouse-core/gather/driver.js (hunk 5 above)
Comment (adamraine) [reply lighthouse-core/gather/driver.js:1131]: Thanks for catching this, this actually explains alot. I thought this PR was working even though there was no LCP lifecycle event, turns out `waitForLcp` just wasn't enabled 😬

**Comment by adamraine** [comment]
Converting this to a draft for now.  Once a lifecycle event or other surface can be used then we can reopen.  The implementation will likely look different from its current state when that happens.

File: lighthouse-core/config/constants.js (hunk 1 above)
Comment (adamraine) [reply lighthouse-core/config/constants.js:57]: What do you think of the exact time value? 40s was basically a guess, would making the default value the same as `maxWaitForLoad` make more sense?

**Comment by connorjclark** [comment]
Let's close until we have a way forward. The PR# is still accessible from the open issue.

**Pull request closed without merging**